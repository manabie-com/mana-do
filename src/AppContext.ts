import { createContext, useContext } from 'react';
import {AppState, initialState} from "./store/reducer";

export type GlobalContextType = {
  state: AppState
  dispatch: Function
}

export const useAppContext = () => useContext<GlobalContextType>(AppContext);

// @ts-ignore
const AppContext = createContext<GlobalContextType>(initialState);

export default AppContext;
