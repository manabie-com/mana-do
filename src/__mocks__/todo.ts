import { TodoStatus } from '@/models/todo'
import { EnhanceTodoStatus } from '@/pages/todo'

export const stateProvider = {
  todos: [
    {
      content: 'Meet some cute peoples at Manabie',
      created_date: '',
      status: TodoStatus.ACTIVE,
      id: '_1',
      user_id: 'aldenn'
    },
    {
      content: 'Work with Manabie',
      created_date: '',
      status: TodoStatus.ACTIVE,
      id: '_2',
      user_id: 'aldenn'
    },
    {
      content: 'Complete the test for todo app',
      created_date: '',
      status: TodoStatus.COMPLETED,
      id: '_3',
      user_id: 'aldenn'
    }
  ]
}

export const statusesFilterProvider: {
  description: string
  status: EnhanceTodoStatus
  numOfRecord: number
}[] = [
  {
    description: 'The number of active todo',
    status: TodoStatus.ACTIVE,
    numOfRecord: 2
  },
  {
    description: 'The number of completed todo',
    status: TodoStatus.COMPLETED,
    numOfRecord: 1
  },
  {
    description: 'All of todo item',
    status: 'ALL',
    numOfRecord: 3
  }
]

export const emptyLabelByStatusProvider: {
  description: string
  status: EnhanceTodoStatus
  label: string
}[] = [
  {
    description:
      "The empty label for ACTIVE status should be Nothing need to do, let's relax",
    status: TodoStatus.ACTIVE,
    label: "Nothing need to do, let's relax"
  },
  {
    description:
      'The empty label for COMPLETED status should be Nothing is completed',
    status: TodoStatus.COMPLETED,
    label: 'Nothing is completed'
  },
  {
    description:
      'The empty label for ALL status should be Type something into the input and Enter',
    status: 'ALL',
    label: 'Type something into the input and Enter'
  }
]
