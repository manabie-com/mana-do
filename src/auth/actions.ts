export const REQUEST = '@Auth/Request'
export const REQUEST_FAIL = '@Auth/Fail'
export const AUTHENTICATE_SUCCESS = '@Auth/AuthenticateSuccess'
export const LOGOUT_SUCCESS = '@Auth/LogoutSuccess'

interface RequestAction {
  type: typeof REQUEST,
}

interface RequestFailAction {
  type: typeof REQUEST_FAIL,
  error: string,
}

interface AuthenticateSuccessAction {
  type: typeof AUTHENTICATE_SUCCESS
  token: string
}

interface LogoutSuccessAction {
  type: typeof LOGOUT_SUCCESS
}

export function request(): RequestAction {
  return { type: REQUEST }
}

export function requestFail(error: string): RequestFailAction {
  return { type: REQUEST_FAIL,error }
}

export function authenticateSuccess(token: string): AuthenticateSuccessAction {
  return { type: AUTHENTICATE_SUCCESS, token }
}

export function logoutSuccess(): LogoutSuccessAction {
  return { type: LOGOUT_SUCCESS }
}

export type Action = RequestAction | RequestFailAction | AuthenticateSuccessAction | LogoutSuccessAction
