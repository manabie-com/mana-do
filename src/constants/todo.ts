export enum TodoStatus {
  ACTIVE = "ACTIVE",
  COMPLETED = "COMPLETED",
}
