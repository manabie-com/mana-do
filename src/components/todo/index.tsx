import React, { useEffect, useReducer, useRef, useState, useMemo, useCallback } from 'react';
import { RouteComponentProps } from 'react-router-dom';

import reducer, { initialState } from 'root/store/reducer';
import {
  setTodos,
  createTodo,
  deleteTodo,
  toggleAllTodos,
  deleteAllTodos,
  updateTodoStatus,
  updateTodoContent
} from 'root/store/actions';
import Service from 'root/service';
import { TodoStatus } from 'root/models/todo';
import { isTodoCompleted } from 'root/utils';
import TodoPresentation from './presentation';

import './_todo.css'

type EnhanceTodoStatus = TodoStatus | 'ALL';


const TodoComponent = ({ history }: RouteComponentProps) => {
  const [{ todos }, dispatch] = useReducer(reducer, initialState);
  const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    (async () => {
      const resp = await Service.getTodos();
      dispatch(setTodos(resp || []));
    })()
  }, [])

  useEffect(() => {
    if (todos.length > 0) {
      const lastItem = document.getElementById(`todo-item-${todos[todos.length - 1].id}`)
      if (lastItem) setTimeout(() => {
        lastItem.scrollIntoView({ block: 'end',  behavior: 'smooth' })
      }, 401) // delay to wait for css animation work first
    }
  }, [todos, showing])

  const onCreateTodo = useCallback(async (e: React.KeyboardEvent<HTMLInputElement>) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    if (e.key === 'Enter' && inputRef.current) {
      try {
        const resp = await Service.createTodo(inputRef.current.value);
        dispatch(createTodo(resp));
        inputRef.current.value = '';
      } catch (e) {
        if (e.response.status === 401) {
          history.push('/')
        }
      }
    }
  }, [dispatch, history])

  const onUpdateTodoStatus: Function = useCallback(async (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
    try {
      await Service.updateTodoStatus(todoId, e.target.checked);
      dispatch(updateTodoStatus(todoId, e.target.checked));
    } catch (error) {

    }
  }, [dispatch])

  const onToggleAllTodo = useCallback(async (e: React.ChangeEvent<HTMLInputElement>) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.toggleAllTodos(e.target.checked)
      dispatch(toggleAllTodos(e.target.checked))
    } catch (error) {

    }
  }, [dispatch])

  const onDeleteAllTodo = useCallback(async () => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.deleteAllTodo()
      dispatch(deleteAllTodos());
    } catch (error) {

    }
  }, [dispatch])

  const handleDeleteTodo = useCallback(async (todoId: string) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.deleteTodo(todoId)
      dispatch(deleteTodo(todoId));
    } catch (error) {

    }
  }, [dispatch])

  const handleUpdateTodoContent = useCallback(async (todoId: string, content: string) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.updateTodoContent(todoId, content)
      dispatch(updateTodoContent(todoId, content));
    } catch (error) {

    }
  }, [dispatch])

  const handleShowingAll = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing('ALL')
  }, [setShowing])

  const handleShowingActive = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing(TodoStatus.ACTIVE)
  }, [setShowing])

  const handleShowingCompleted = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing(TodoStatus.COMPLETED)
  }, [setShowing])

  const showTodos = useMemo(() => todos.filter((todo) => {
    /**
     * Reduce recalculating showing todos, Does not recalculate showing todos when todos and showing state have not changed
     */
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      case 'ALL':
      default:
        return true;
    }
  }), [showing, todos])

  const activeTodos = useMemo(() => todos.reduce(function (accum, todo) {
    /**
     * Reduce recalculating activeTodos, Does not recalculate active todos when todos has not changed
    */
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0), [todos])

  return (
    <TodoPresentation
      inputRef={inputRef}
      onCreateTodo={onCreateTodo}
      showTodos={showTodos}
      onUpdateTodoStatus={onUpdateTodoStatus}
      deleteTodo={handleDeleteTodo}
      todos={todos}
      activeTodos={activeTodos}
      onToggleAllTodo={onToggleAllTodo}
      handleShowingAll={handleShowingAll}
      handleShowingActive={handleShowingActive}
      handleShowingCompleted={handleShowingCompleted}
      onDeleteAllTodo={onDeleteAllTodo}
      onUpdateTodoContent={handleUpdateTodoContent}
      showing={showing}
    />
  );
};

export default TodoComponent;