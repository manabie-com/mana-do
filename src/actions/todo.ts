import { Todo } from "../models/todo";
import { ActionTypes } from "../constants";
import { Action } from "./commons";

export function setTodos(todos: Array<Todo>): Action<Array<Todo>> {
    return {
        type: ActionTypes.SET_TODO,
        payload: todos
    }
}


export function createTodo(newTodo: Todo): Action<Todo> {
    return {
        type: ActionTypes.CREATE_TODO,
        payload: newTodo
    }
}

export function updateTodoStatus(todoId: string, checked: boolean): Action<{ todoId: string, checked: boolean }> {
    return {
        type: ActionTypes.UPDATE_TODO_STATUS,
        payload: {
            todoId,
            checked
        }
    }
}

export function deleteTodo(todoId: string): Action<string> {
    return {
        type: ActionTypes.DELETE_TODO,
        payload: todoId
    }
}


export function deleteAllTodos(): Action<never> {
    return {
        type: ActionTypes.DELETE_ALL_TODOS,
    }
}

export function toggleAllTodos(checked: boolean): Action<boolean> {
    return {
        type: ActionTypes.TOGGLE_ALL_TODOS,
        payload: checked
    }
}