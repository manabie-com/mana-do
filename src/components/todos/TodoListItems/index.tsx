import React, { useState } from "react";
import { findTodoById, findTodoIndexById, updateTodoContent } from "../../../utils/todoUtils";
import { isTodoCompleted } from "../../../utils";
import { deleteTodo, updateTodoContent as updateTodoContentAction, updateTodoStatus } from "../../../store/actions";
import { getTodosFromStorage } from "../../../utils/storageUtils";
import { useAppContext } from "../../../AppContext";
import { ListItemWrapper } from "./styles";
import { Draggable } from "react-beautiful-dnd";
import CrossIcon from "../../../assets/icons/CrossIcon";
import { Todo } from "../../../models/todo";

interface TodoListItemsProps {
  id: string;
  index?: number;
}

const TodoListItems = ({ id, index = 0 }: TodoListItemsProps) => {
  const {
    state: { todos },
    dispatch,
  } = useAppContext();
  const [todosValue, setTodosValue] = useState(todos);
  const todo = findTodoById(todos, id) ?? ({} as Todo);
  const [toggle, setToggle] = useState({});
  const foundIndex = findTodoIndexById(todosValue, id);
  const text = todosValue[foundIndex].content;

  const onUpdateTodoStatus = (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
    dispatch(updateTodoStatus(todoId, e.target.checked));
  };

  const deleteTodoItem = (todoId: string) => {
    dispatch(deleteTodo(todoId));
  };

  const onDoubleClickItem = (e: React.MouseEvent<HTMLDivElement>) => {
    setToggle((prev) => ({
      ...prev,
      [id]: true,
    }));
  };

  const onLiveEditKeydown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    const target = event.target as HTMLTextAreaElement;
    if (event.key === "Enter" || event.key === "Escape") {
      if (event.key === "Escape") {
        setTodosValue(getTodosFromStorage());
      }

      if (event.key === "Enter") {
        const newTodos = updateTodoContent(todosValue, todo.id, target.value);
        setTodosValue(newTodos);
        dispatch(updateTodoContentAction(todo.id, target.value));
      }

      setToggle((prev) => ({
        ...prev,
        [id]: false,
      }));
    }
  };

  const onChangeLiveEdit = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    const foundIndex = findTodoIndexById(todos, todo.id);
    let newTodos = [...todos];
    newTodos[foundIndex] = {
      ...newTodos[foundIndex],
      content: value,
    };
    setTodosValue(newTodos);
  };

  const onBlurLiveEdit = () => {
    setToggle((prev) => ({
      ...prev,
      [id]: false,
    }));
    setTodosValue(getTodosFromStorage());
  };

  return (
    <Draggable key={id} draggableId={"draggable-" + id} index={index}>
      {(provided, snapshot) => (
        <ListItemWrapper
          key={id}
          className="list-item"
          onDoubleClick={onDoubleClickItem}
          ref={provided.innerRef}
          {...provided.draggableProps}
          {...provided.dragHandleProps}
          data-testid="list-item-wrapper"
        >
          <div className="text_and_checkbox">
            <input
              id={todo.id}
              type="checkbox"
              data-testid="toggle-all"
              checked={isTodoCompleted(todo)}
              onChange={(e) => onUpdateTodoStatus(e, todo.id)}
            />
            {!toggle[id] && <label htmlFor={todo.id}>{text}</label>}
            {toggle[id] && (
              <input
                autoFocus
                type="text"
                className="input_live_edit"
                value={text}
                onChange={onChangeLiveEdit}
                onBlur={onBlurLiveEdit}
                onKeyDown={onLiveEditKeydown}
                data-testid="live-edit-todo-content"
              />
            )}
          </div>

          <div className="delete">
            <button className="delete-btn" data-testid="delete-all-todo-item" onClick={() => deleteTodoItem(id)}>
              <CrossIcon />
            </button>
          </div>
        </ListItemWrapper>
      )}
    </Draggable>
  );
};

export default TodoListItems;
