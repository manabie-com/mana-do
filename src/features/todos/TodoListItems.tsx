import React, {useRef, useState} from 'react';
import {FaTrash} from "react-icons/fa";

import {findTodoById, findTodoIndexById, updateTodoContent} from '../../utils/todoUtils';
import {GlobalContextType, StateContext} from "../../App";
import {isTodoCompleted} from "../../utils";
import {deleteTodo, updateTodoContent as updateTodoContentAction, updateTodoStatus} from "../../store/actions";
import {getTodosFromStorage} from "../../utils/storeageUtils";

interface TodoListItemsProps {
  id: string;
}

const TodoListItems = ({id}: TodoListItemsProps) => {
  const {state: {todos}, dispatch} = React.useContext<GlobalContextType>(StateContext);
  const [todosValue, setTodosValue] = useState(todos);
  const listRef = useRef(null);
  const todoItemRef = useRef<Array<HTMLInputElement | null>>([]);
  const todo = findTodoById(todos, id) ?? {id: '', user_id: '', content: '', created_date: ''};
  const [toggle, setToggle] = useState({});

  const onUpdateTodoStatus = (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
    dispatch(updateTodoStatus(todoId, e.target.checked));
  }

  const onDeleteTodoItem = (todoId: string) => {
    if (listRef && listRef.current) {
      // @ts-ignore
      listRef.current.classList.add('hidden');
    }

    const t = window.setTimeout(() => {
      window.clearTimeout(t);
      dispatch(deleteTodo(todoId));
    }, 300);
  }

  const foundIndex = findTodoIndexById(todosValue, id);

  const onLiveEditKeydown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    const target = event.target as HTMLTextAreaElement;
    if (event.key === 'Enter' || event.key === 'Escape') {
      if (event.key === 'Escape') {
        setTodosValue(getTodosFromStorage());
      }

      if (event.key === 'Enter') {
        const newTodos = updateTodoContent(todosValue, todo.id, target.value)
        setTodosValue(newTodos);
        dispatch(updateTodoContentAction(todo.id, target.value))
      }

      setToggle((prev) => ({
        ...prev,
        [id]: false
      }));
    }
  }

  const onChangeLiveEdit = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    const foundIndex = findTodoIndexById(todos, todo.id);
    let newTodos = [...todos];
    newTodos[foundIndex] = {
      ...newTodos[foundIndex],
      content: value
    }
    setTodosValue(newTodos);
  }

  const onBlurLiveEdit = () => {
    setToggle((prev) => ({
      ...prev,
      [id]: false
    }));
    setTodosValue(getTodosFromStorage());
  }

  return (
    <li ref={listRef} className={`ToDo__item visible ${toggle[id] ? 'border-blue' : ''}`}>
      <input
        type="checkbox"
        checked={isTodoCompleted(todo)}
        onChange={(e) => onUpdateTodoStatus(e, todo.id)}
      />
      <span className="Todo__content_wrap" onDoubleClick={() => {
        setToggle((prev) => ({
          ...prev,
          [id]: true
        }));
        todoItemRef.current[id]?.focus();
      }}>
        <input
          style={{left: toggle[id] ? '0' : '-9999px'}}
          type="text"
          className="Todo__content_live_edit"
          value={todosValue[foundIndex].content}
          ref={el => todoItemRef.current[id] = el}
          onChange={onChangeLiveEdit}
          onBlur={onBlurLiveEdit}
          onKeyDown={onLiveEditKeydown}
        />
        <span
          style={{display: toggle[id] ? 'none' : 'inline-block'}}>
          {todosValue[foundIndex].content}
        </span>
      </span>
      <button
        className="Todo__delete"
        onClick={() => onDeleteTodoItem(id)}
      >
        <FaTrash/>
      </button>
    </li>
  )
}

export default TodoListItems;
