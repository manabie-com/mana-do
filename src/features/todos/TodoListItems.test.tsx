import React from 'react';
import renderer from "react-test-renderer";
import {shallow} from "enzyme";

import TodoListItems from "./TodoListItems";
import * as AppContext from "../../AppContext";
import {TodoStatus} from "../../models/todo";
import * as actions from "../../store/actions";
import {storeReducerStorage} from "../../utils/storeageUtils";


jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useRef: jest.fn(),
  useState: jest.fn(),
}))

const todoId = 'hfuwyr';

const mockedTodos = [
  {
    id: todoId,
    content: 'todo 1',
    user_id: 'firstUser',
    status: TodoStatus.All,
    created_date: ''
  },
  {
    id: 'orislaj',
    content: 'todo 2',
    user_id: 'firstUser',
    status: TodoStatus.All,
    created_date: ''
  },
  {
    id: 'jiwuro',
    content: 'todo 3',
    user_id: 'firstUser',
    status: TodoStatus.Active,
    created_date: ''
  },
  {
    id: 'iejals',
    content: 'todo 4',
    user_id: 'firstUser',
    status: TodoStatus.Active,
    created_date: ''
  },
  {
    id: 'jfuqos',
    content: 'todo 5',
    user_id: 'firstUser',
    status: TodoStatus.Completed,
    created_date: ''
  },
  {
    id: 'ahsyel',
    content: 'todo 6',
    user_id: 'firstUser',
    status: TodoStatus.Completed,
    created_date: ''
  },
];

beforeEach(() => {
  const mockContextValue = {
    state: {
      todos: mockedTodos,
    },
    dispatch: jest.fn()
  }

  jest
    .spyOn(AppContext, 'useAppContext')
    // @ts-ignore
    .mockImplementation(() => mockContextValue);
})

describe('snapshot', () => {
  test('should return snapshot', () => {
    const tree = renderer.create(<TodoListItems id={todoId}/>).toJSON();
    expect(tree).toMatchSnapshot();
  })
})

describe('handle events', () => {

  afterEach(() => {
    jest.clearAllMocks();
  })

  const mockClassListAddMethod = jest.fn();
  const setStateMock = jest.fn();
  const useStateMock = (initialValue: any) => [initialValue, setStateMock];
  const useRefMock = () => ({
    current: {
      classList: {
        add: mockClassListAddMethod
      }
    }
  });
  jest.spyOn(React, 'useRef').mockImplementation(useRefMock);
  // @ts-ignore
  jest.spyOn(React, 'useState').mockImplementation(useStateMock);

  test('should update todo status after click on toggle all checkbox', () => {
    const updateTodoStatus = jest.spyOn(actions, 'updateTodoStatus');
    const event = {
      target: {
        checked: true
      }
    }
    const wrapper = shallow(<TodoListItems id={todoId}/>);
    const toggleCheckbox = wrapper.find({'data-testid': 'toggle-all'});
    const toggleCheckboxProps = toggleCheckbox.props();

    // handle on toggle
    toggleCheckboxProps.onChange(event, todoId);
    expect(updateTodoStatus).toHaveBeenCalledWith(todoId, event.target.checked)
  });

  test('should fade-out and delete a todo item by id', () => {
    jest.useFakeTimers();

    const deleteTodo = jest.spyOn(actions, 'deleteTodo');
    const wrapper = shallow(<TodoListItems id={todoId}/>);
    const deleteButton = wrapper.find({'data-testid': 'delete-all-todo-item'});
    const deleteButtonProps = deleteButton.props();

    // handle on click
    deleteButtonProps.onClick();

    jest.runAllTimers();

    expect(mockClassListAddMethod).toHaveBeenCalledWith('hidden');
    expect(deleteTodo).toHaveBeenCalledWith(todoId);
  });

  test('should update todo status live by Enter keyboard', () => {
    const event = {
      key: 'Enter',
      target: {
        value: 'todo item edited'
      }
    }

    const expectedTodos = [
      {
        id: todoId,
        content: event.target.value,
        user_id: 'firstUser',
        status: TodoStatus.All,
        created_date: ''
      },
      {
        id: 'orislaj',
        content: 'todo 2',
        user_id: 'firstUser',
        status: TodoStatus.All,
        created_date: ''
      },
      {
        id: 'jiwuro',
        content: 'todo 3',
        user_id: 'firstUser',
        status: TodoStatus.Active,
        created_date: ''
      },
      {
        id: 'iejals',
        content: 'todo 4',
        user_id: 'firstUser',
        status: TodoStatus.Active,
        created_date: ''
      },
      {
        id: 'jfuqos',
        content: 'todo 5',
        user_id: 'firstUser',
        status: TodoStatus.Completed,
        created_date: ''
      },
      {
        id: 'ahsyel',
        content: 'todo 6',
        user_id: 'firstUser',
        status: TodoStatus.Completed,
        created_date: ''
      },
    ]
    const updateTodoContent = jest.spyOn(actions, 'updateTodoContent');
    const wrapper = shallow(<TodoListItems id={todoId}/>);
    const inputEditTodoContent = wrapper.find({'data-testid': 'live-edit-todo-content'});
    const inputEditTodoContentProps = inputEditTodoContent.props();

    // handle on change
    inputEditTodoContentProps.onChange(event);
    expect(setStateMock).toHaveBeenCalledWith(expectedTodos);

    // handle on Enter keydown
    inputEditTodoContentProps.onKeyDown(event);
    expect(setStateMock).toHaveBeenCalledWith(expectedTodos);
    expect(updateTodoContent).toHaveBeenCalledWith(todoId, event.target.value);
  });

  test('should discard todo status live by Escape keyboard', () => {
    const event = {
      key: 'Escape',
      target: {
        value: 'todo item edited'
      }
    }

    const expectedTodos = [
      {
        id: todoId,
        content: event.target.value,
        user_id: 'firstUser',
        status: TodoStatus.All,
        created_date: ''
      },
      {
        id: 'orislaj',
        content: 'todo 2',
        user_id: 'firstUser',
        status: TodoStatus.All,
        created_date: ''
      },
      {
        id: 'jiwuro',
        content: 'todo 3',
        user_id: 'firstUser',
        status: TodoStatus.Active,
        created_date: ''
      },
      {
        id: 'iejals',
        content: 'todo 4',
        user_id: 'firstUser',
        status: TodoStatus.Active,
        created_date: ''
      },
      {
        id: 'jfuqos',
        content: 'todo 5',
        user_id: 'firstUser',
        status: TodoStatus.Completed,
        created_date: ''
      },
      {
        id: 'ahsyel',
        content: 'todo 6',
        user_id: 'firstUser',
        status: TodoStatus.Completed,
        created_date: ''
      },
    ]
    const data = {
      todos: mockedTodos
    }
    // Handle Set local storage
    storeReducerStorage(data);
    const wrapper = shallow(<TodoListItems id={todoId}/>);
    const inputEditTodoContent = wrapper.find({'data-testid': 'live-edit-todo-content'});
    const inputEditTodoContentProps = inputEditTodoContent.props();

    // handle on change
    inputEditTodoContentProps.onChange(event);
    expect(setStateMock).toHaveBeenCalledWith(expectedTodos);

    // handle on Enter keydown
    inputEditTodoContentProps.onKeyDown(event);
    expect(setStateMock).toHaveBeenCalledWith(data.todos);
  });

  // TODO: discard todo status live by onblur
  // test('should discard todo status live by onblur', () => {
  //
  // });
})
