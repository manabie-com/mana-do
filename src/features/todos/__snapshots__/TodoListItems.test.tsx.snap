// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should return snapshot 1`] = `
<li
  className="ToDo__item visible "
>
  <input
    checked={false}
    data-testid="toggle-all"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="Todo__content_wrap"
    onDoubleClick={[Function]}
  >
    <input
      className="Todo__content_live_edit"
      data-testid="live-edit-todo-content"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      style={
        Object {
          "left": "-9999px",
        }
      }
      type="text"
      value="todo 1"
    />
    <span
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      todo 1
    </span>
  </span>
  <button
    className="Todo__delete"
    data-testid="delete-all-todo-item"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
      />
    </svg>
  </button>
</li>
`;
